File: _engine/typescript/v0.1.0/request.ts
// Copyright 2026 GildenEye
// SPDX-License-Identifier: Apache-2.0

import type { ModelRequestParams, ModelResponse, AIFuncConfig } from './types';

const ENGINE_DEFAULT_TIMEOUT = 7000;
const ENGINE_DEFAULT_MAX_RETRIES = 1;

/** Subset of aifunc.json fields injected at build time by the CLI. */
export interface ProjectDefaults {
  timeout?: number;
  maxRetries?: number;
}

export async function sendRequest(
  config: AIFuncConfig,
  params: ModelRequestParams,
  projectDefaults: ProjectDefaults = {}
): Promise<ModelResponse> {
  if (!config.baseURL) {
    throw new Error('AIFuncConfig.baseURL is required when mock mode is disabled');
  }
  if (!config.apiKey) {
    throw new Error('AIFuncConfig.apiKey is required when mock mode is disabled');
  }

  const base = config.baseURL.replace(/\/$/, '');
  const url = base.endsWith('/chat/completions')
    ? base
    : `${base}/chat/completions`;

  // Priority: user config > aifunc.json (projectDefaults) > engine default
  const timeout = config.timeout ?? projectDefaults.timeout ?? ENGINE_DEFAULT_TIMEOUT;
  const maxRetries = config.maxRetries ?? projectDefaults.maxRetries ?? ENGINE_DEFAULT_MAX_RETRIES;

  let lastError: Error = new Error('Unknown error during model request');

  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      return await _doRequest(url, config.apiKey, params, timeout);
    } catch (err) {
      lastError = err instanceof Error ? err : new Error(String(err));
      if (attempt < maxRetries) {
        continue;
      }
    }
  }

  throw lastError;
}

async function _doRequest(
  url: string,
  apiKey: string,
  params: ModelRequestParams,
  timeout: number
): Promise<ModelResponse> {
  const controller = new AbortController();
  const timeoutId = setTimeout(() => controller.abort(), timeout);

  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${apiKey}`,
      },
      body: JSON.stringify(params),
      signal: controller.signal,
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`Model API returned ${response.status}: ${errorText.slice(0, 500)}`);
    }

    return await response.json() as ModelResponse;
  } catch (error) {
    if (error instanceof Error) {
      if (error.name === 'AbortError') {
        throw new Error(`Request timeout after ${timeout}ms`);
      }
      throw error;
    }
    throw new Error('Unknown error during model request');
  } finally {
    clearTimeout(timeoutId);
  }
}

File: _engine/go/v0.1.0/request.go
// Copyright 2026 GildenEye
// SPDX-License-Identifier: Apache-2.0

package engine

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
)

// defaultHTTPClient has no timeout; per-request context handles cancellation.
var defaultHTTPClient = &http.Client{}

const maxBodySize = 64 * 1024 * 1024

func SendRequest(ctx context.Context, config AIFuncConfig, params ModelRequestParams) (ModelResponse, error) {
	if config.BaseURL == "" {
		return ModelResponse{}, fmt.Errorf("AIFuncConfig.BaseURL is required when mock mode is disabled")
	}
	if config.APIKey == "" {
		return ModelResponse{}, fmt.Errorf("AIFuncConfig.APIKey is required when mock mode is disabled")
	}

	base := strings.TrimRight(config.BaseURL, "/")
	endpoint := base
	if !strings.HasSuffix(base, "/chat/completions") {
		endpoint = base + "/chat/completions"
	}

	bodyBytes, err := json.Marshal(params)
	if err != nil {
		return ModelResponse{}, fmt.Errorf("marshal request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, bytes.NewReader(bodyBytes))
	if err != nil {
		return ModelResponse{}, fmt.Errorf("build request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+config.APIKey)

	client := config.HTTPClient
	if client == nil {
		client = defaultHTTPClient
	}
	resp, err := client.Do(req)
	if err != nil {
		return ModelResponse{}, err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		preview, _ := io.ReadAll(io.LimitReader(resp.Body, 500))
		return ModelResponse{}, fmt.Errorf("model API returned %d: %s", resp.StatusCode, string(preview))
	}

	respBody, err := io.ReadAll(io.LimitReader(resp.Body, maxBodySize))
	if err != nil {
		return ModelResponse{}, fmt.Errorf("read response body: %w", err)
	}

	var modelResp ModelResponse
	if err := json.Unmarshal(respBody, &modelResp); err != nil {
		return ModelResponse{}, fmt.Errorf("failed to parse model response: %w", err)
	}
	return modelResp, nil
}
